### PR TITLE
use Luhn to calculate ar_SA id numbers.

### PR DESCRIPTION
**Pull request by FooBarQuaxx**

### CODE REVIEW

File: src/Faker/Provider/ar_SA/Company.php [hunk 1]
```diff
@@ -60,4 +62,16 @@ public function bs()
...
         return join($result, ' ');
     }
+
+    /**
+     * example 7001010101
+     **/
+    public static function companyIdNumber()
+    {
+        do {
+            $number = static::numerify('700#######');
```
Comment (fzaninotto) [inline src/Faker/Provider/ar_SA/Company.php]: nah, you can do better than try until it's valid!

Comment (FooBarQuaxx) [inline src/Faker/Provider/ar_SA/Company.php]: Actually this is exactly what is done on the Person provider for the same locale, but I will look for something.

Comment (okj579) [inline src/Faker/Provider/ar_SA/Company.php]: `Luhn::computeCheckDigit()` will give you the correct last digit.

Comment (fzaninotto) [inline src/Faker/Provider/ar_SA/Company.php]: Then the Person provider is wrong, too, and you can update it while you're touching Luhns.

Comment (FooBarQuaxx) [inline src/Faker/Provider/ar_SA/Company.php]: I think I did on commit 757750f.

Comment (fzaninotto) [inline src/Faker/Provider/ar_SA/Company.php]: It doesn't appear in this PR

Comment (FooBarQuaxx) [inline src/Faker/Provider/ar_SA/Company.php]: I don't want to sound stupid, but I see the changes on the last commit on this PR.

File: src/Faker/Provider/ar_SA/Utils.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Faker\Provider\ar_SA;
+
+class Utils
+{
+    // ripped unashamedly from https://github.com/grahamking/darkcoding-credit-card/blob/master/gencc.php
+    public static function generateLuhnNumber($prefix, $length)
```
Comment (fzaninotto) [inline src/Faker/Provider/ar_SA/Utils.php]: You already have everything in `Faker\Calculator\Luhn`. No need to reimplement it here.

Comment (FooBarQuaxx) [inline src/Faker/Provider/ar_SA/Utils.php]: Is it good enough now?

File: src/Faker/Provider/ar_SA/Utils.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Faker\Provider\ar_SA;
+
+use Faker\Calculator\Luhn;
+use Faker\Provider\Base as BaseProvider;
+
+class Utils
+{
+    public static function generateLuhnNumber($prefix, $length)
```
Comment (fzaninotto) [inline src/Faker/Provider/ar_SA/Utils.php]: this would be better placed in `Faker\Calculator\Luhn`

**Comment by FooBarQuaxx** [comment]
Anything new about this pull request?

File: src/Faker/Calculator/Luhn.php [hunk 4]
```diff
@@ -55,4 +58,41 @@ public static function isValid($number)
...
+        if (strlen($prefix) == $length) {
+            if (static::isValid($prefix)) {
+                return $prefix;
+            } else {
+                throw new InvalidArgumentException('prefix length equals length but prefix is not a Luhn number.');
+            }
+        }
+
+        $pattern = $prefix . str_repeat('#', $length - strlen($prefix) - 1);
+        $partialValue = BaseProvider::numerify($pattern);
```
Comment (fzaninotto) [inline src/Faker/Calculator/Luhn.php]: you should pass the partialValue as parameter to this method, to separate the responsibilities: Luhn calculation (this class), random generation (Company.php)

**Comment by FooBarQuaxx** [comment]
I have made the changes as I understand how they should be done. Please check.

**Comment by fzaninotto** [comment]
Excellent, thanks for your contribution!

**Comment by FooBarQuaxx** [comment]
Thank you too for help me through it.

**Pull request merged**